CredentialProvider.Microsoft/CredentialProviders/VstsBuildTaskServiceEndpoint/VstsBuildTaskMsalTokenProvidersFactory.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Artifacts.Authentication;
+using Microsoft.Extensions.Logging;
+using NuGetCredentialProvider.Util;
+
+namespace NuGetCredentialProvider.CredentialProviders.VstsBuildTaskServiceEndpoint;
+
+internal class VstsBuildTaskMsalTokenProvidersFactory : ITokenProvidersFactory
+{
+    private readonly ILogger logger;
+
+    public VstsBuildTaskMsalTokenProvidersFactory(ILogger logger)
+    {
+        this.logger = logger;
+    }
+
+    public Task<IEnumerable<ITokenProvider>> Get(Uri authority)
+    {
+        var app = AzureArtifacts.CreateDefaultBuilder(authority)
+            .WithBroker(EnvUtil.MsalAllowBrokerEnabled(), logger)
+            .WithHttpClientFactory(HttpClientFactory.Default)
+            .WithLogging(
+                (level, message, containsPii) =>
+                {
+                    // We ignore containsPii param because we are passing in enablePiiLogging below.
+                    logger.LogTrace("MSAL Log ({level}): {message}", level, message);
+                },
+                enablePiiLogging: EnvUtil.GetLogPIIEnabled()
+            )
+            .Build();
+
+        return Task.FromResult(MsalTokenProviders.Get(app, logger)
+            .Where(x => x.Name == "MSAL Managed Identity" || x.Name == "MSAL Service Principal"));
```

Comment:
Just construct the two we need here instead of filtering... think this means we can avoid creating the app entirely above?

Reply:
I would still need to create the app above since the two we need to construct take in the app as a constructor parameter. Additionally, the ITokenProviderFactory interface is not super friendly for the proposed scenario since it is returning a Task<IEnumerable<ITokenProvider>>

Reply:
cache should be added which should help a bit since the async was required for that.. then it's just returning an IEnumerable

Reply:
I can add it back but I initially removed it due to: https://learn.microsoft.com/en-us/entra/msal/java/advanced/managed-identity#caching not being supported

Reply:
but service principals it should work?

Reply:
actually, the cache needs to be registered on the app which you're creating inside the token providers now. so either no cache, or create the apps on the outside and pass them in

Reply:
removed - I think this is a much larger change in the [vstsbuildendpoint cred as well](https://github.com/microsoft/artifacts-credprovider/blob/74fe273afe1270081d69cc219133afd0d230fba6/CredentialProvider.Microsoft/CredentialProviders/VstsBuildTaskServiceEndpoint/VstsBuildTaskServiceEndpointCredentialProvider.cs#L48). to support caching for only service principal returned tokens.